eng/common/scripts/get-codeowners/get-codeowners.ps1

```diff
@@ -0,0 +1,84 @@
+<#
```

Comment:
I assume you moved this to try to align the directory structure with the tests. So far we haven't been splitting the scripts up unless there is a group of them. So if there isn't a strong reason I would leave it in the root of the scripts directory.

Reply:
Oh I see you also added the helper functions script, for some reason I thought you put that under Helpers. I would suggest you rename that to codeowners-helpers.ps1 and put that under Helpers and then move the get-codeowners.ps1 back to the root of the scripts folder.

Reply:
@weshaggard I was following the job-matrix convention:

![image](https://user-images.githubusercontent.com/4429827/222795835-61b52b3d-26c1-4800-8d9a-ce7222279fd8.png)

I also have a strong dislike for the name `helper`, as a concept that is too vague and hence lacking cohesion. It is akin to creating a C# class named `Methods` that has all the methods used anywhere in the code. Or a class called `Constants` that has all the constants.

Now the `get-codeowners` dir contains two files that form a part of cohesive whole: the entry point of `get-codeowners.ps1` and the internal functions it leverages, `get-codeowners-functions.ps1`. The fact `get-codeowners.ps1` is is its own dir it also means I can point to that dir from the `codeowners-parser` `ci.yml` and run only tests relevant to it, instead of all tests.

Reply:
I'm not hung up on the word helpers but the goal is to eventually have the entry point scripts live in the root and have some set of helper functions be separated from the entry point scripts so that they can be like a set of shared code that could be imported by any of the scripts. I would like to make it clear which set of scripts you call and which are used by other scripts. 

If we want to use a "-functions" convention I'd be OK with that as well, but I'm not sure it is worth having a subdirectory for every entry-point script. For larger areas I think it could be worth it. 

@benbp I'd be interested in your thoughts here as well.

Reply:
I see your point on "scripts dir has all the scripts that are valid entry points" but I want to clarify that `get-codeowners-functions.ps1` is not be used by anything else except these two use cases:
- to be called by `get-codeowners.ps1`
- to be called by `common-tests/get-codeowners/get-codeowenrs.tests.ps1` Pester unit tests 

Basically the `get-codeowners.ps1` functions is equivalent to C# `internal` class methods that have the `[InternalsVisibleTo]` attribute to enable the unit test project (here: Pester) to call them directly to simplify the UT logic, as calling `get-codeowners.ps1` directly from the Pester unit tests would be cumbersome.

Reply:
This is the pester usage:

``` powershell
. $PSScriptRoot/../../common/scripts/get-codeowners/get-codeowners-functions.ps1

function TestGetCodeowners([string]$targetPath, [string]$codeownersFileLocation, [bool]$includeNonUserAliases = $false, [string[]]$expectedOwners) {
  Write-Host "Test: Owners for path '$targetPath' in CODEOWNERS file at path '$codeownersFileLocation' should be '$expectedOwners'"
  
  $actualOwners = Get-Codeowners `
    -ToolVersion $ToolVersion `
    -ToolPath $ToolPath `
    -DevOpsFeed $DevOpsFeed `
    -VsoVariable $VsoVariable `
    -targetPath $targetPath `
    -codeownersFileLocation $codeownersFileLocation `
    -includeNonUserAliases $IncludeNonUserAliases

  $actualOwners.Count | Should -Be $expectedOwners.Count
  for ($i = 0; $i -lt $expectedOwners.Count; $i++) {
    $expectedOwners[$i] | Should -Be $actualOwners[$i]
  }
}
```

and then:
``` powershell
      $resolvedCodeownersPath = (Resolve-Path $codeownersPath)
      TestGetCodeowners `
        -targetPath $targetPath `
        -codeownersFileLocation $resolvedCodeownersPath `
        -includeNonUserAliases $true `
        -expectedOwners $expectedOwners
```
`

Reply:
Does the script need to be split up to be tested in pester? As long as the script is split into functions you should be able to `.` source the entry point script as well.

Reply:
@weshaggard from what I see dot-sourcing is executing the top-level logic [1]. To my understanding we want for the scripts in `eng/scripts` dir be executable by just invoking them (instead of dot-sourcing and then calling a function) hence they need to have top-level logic, hence we cannot follow the "Dot-source in BeforeAll" pattern in Pester tests, unless we are dot-sourcing a ps1 file that has no top-level logic, like the `-functions.ps1` file I introduced.

[1]

Consider `test.ps1`:

``` powershell
function Foo()
{
  Write-Host "Hello from Foo!"
}

Write-Host "Hello from root!"
```

And following session snippets:
``` powershell
. ./test.ps1
Hello from root!
```

Reply:
@weshaggard I guess I could scrap dot-sourcing in `BeforeAll` and invoke the `get-codeowners.ps1` directly from the test case. It is effectively a pass-through delegate to the dot-sourced `Get-Codeowners` function anyway (see snippet above). This way I could avoid introducing the `-functions.ps1` file. But I have a suspicion this won't always be the case, i.e. sometimes we really would prefer skipping the top-level logic and testing one of the functions directly. If so, then we would end up with inconsistent approach in our tests: sometimes dot-sourcing in BeforeAll (to skip the top-level logic), sometimes calling directly. I feel we should avoid such inconsistency.

I am also going to ask @benbp what is his take on this :)

Reply:
I started introducing the `-functions.ps1` or `-lib.ps1` pattern to function as a sort of internal class as @konrad-jamrozik described, so that we could easily call things from test, dot source functions locally for development and bug repros in the shell, and nudge towards better code organization (many of our scripts which haven't done this have grown to hundreds of lines of untestable top-level spaghetti over the years). I also agree that dot sourcing and invoking functions directly is a better pattern for unit tests.

To me this is distinctly different from `-Helpers.ps1` files for which inner functions are intended to be shared more broadly, and the script itself isn't a cohesive whole but rather a grab bag of related pieces of functionality.

There is a way to avoid the whole dot source/direct invocation difficulties via [this directive](https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/scripts/stress-testing/generate-scenario-matrix.ps1#L90):

```
if ($MyInvocation.InvocationName -ne ".") {
    # Call entrypoint function
}
```

However we ran into annoyances in some devops contexts, for example if you are calling your script within the azure cli task to make use of login credentials then it actually ends up dot sourcing the script, so nothing is invoked. Due to this and the code organization considerations I listed above, I have tried to use the lib/functions file and invocation file convention instead.

I agree with @weshaggard that if we just have just two files it's probably not worth its own directory.

Reply:
Re test convention around dot sourcing and calling directly, I think this boils down to two fundamental approaches to powershell development:

1. Use files as the de-facto function boundary, defining one cmdlet per file (with optional micro-helper functions inside). I would say this is a little bit more in line with the powershell language, in that it's easier to consume and define behavior for built-ins like global/script scopes, common variables like `$ErrorActionPreference` and common cmdlet parameters like `-WhatIf`.
2. Use functions as the function boundary. It can be a little more annoying to keep various globally scoped settings and values from colliding, but I think it creates a better development experience (not having to traverse dozens of files, not having to maintain dot source hierarchies).

Like I said in my above comment, I try to trend towards approach 2, because with approach 1 it's hard to keep people from adding too much code over time to what are intended to be small files. Whereas if one is editing a file where all code is in functions, naturally they are going to add their own code into a function.

Back to the testing perspective, I think it's ok to have tests that run the whole file, and tests that run functions from within the file. However my ideal pattern is that the lib/functions file will have a main/entrypoint function defined, and if the test wants to call the whole flow, it dot sources the file and calls main, rather than invoking the script file. This gives us some flexibility with how we allow users to invoke scripts without the unit tests having to follow the same format.

Reply:
@benbp @weshaggard OK so I think we mostly agree that `-functions.ps1` is a good idea.

Re this:

> I agree with @weshaggard that if we just have just two files it's probably not worth its own directory.

I want to point out that [job-matrix](https://github.com/Azure/azure-sdk-tools/tree/main/eng/common/scripts/job-matrix) has its own directory. Plus if we move `-functions.ps1` to `/eng/common/scripts/`, then we will break the rule that "all top-level `.ps1` files are runnable". Hence I would still keep `/eng/common/scripts/get-codeowners/`.

Reply:
I'm good with most of is being said here but we had a little further discussion in the teams channel. Mostly whatever we decide we should document in our powershell.md file and start to move towards that convention.

Reply:
@weshaggard OK! How about I merge this PR, then make a PR updating the conventions doc, and also updating the layout to match the conventions? I want to get this PR in as it also has some other changes besides file layout, including changes to `eng/common`, so I don't want to make it linger.

Reply:
Sounds good.